[Response Ops][Docs] Alerting circuit breaker docs (#131459)
elastic/kibana: docs/user/production-considerations/alerting-production-considerations.asciidoc

For more information on index lifecycle management, see:
{ref}/index-lifecycle-management.html[Index Lifecycle Policies].

[float]
[[alerting-circuit-breakers]]
=== Circuit breakers

> There are several scenarios where running alerting rules and actions can start to negatively impact the overall health of a {kib} instance either by clogging up Task Manager throughput or by consuming so much CPU/memory that other operations cannot complete in a reasonable amount of time. We have added several <<alert-settings,configurable>> circuit breakers to help minimize these effects.

Reviewer: We generally avoid "we" except in release highlights or tutorials:
```suggestion
There are several scenarios where running alerting rules and actions can start to negatively impact the overall health of a {kib} instance either by clogging up Task Manager throughput or by consuming so much CPU/memory that other operations cannot complete in a reasonable amount of time. There are several <<alert-settings,configurable>> circuit breakers to help minimize these effects.
```

[float]
==== Rules with very short intervals

> Running large numbers of rules at very short intervals can quickly clog up Task Manager throughput, leading to higher schedule drift. Use `xpack.alerting.rules.minimumScheduleInterval.value` to set a minimum schedule interval for rules. The default (and recommended) value for this configuration is `1m`. Use `xpack.alerting.rules.minimumScheduleInterval.enforce` to specify whether to strictly enforce this minimum. While the default value for this is `false` to maintain backwards compatibility with existing rules, we recommend setting this to `true` to prevent new and updated rules from running at an interval below the minimum.

Reviewer: I suggest dropping the "we" and either using "set this to ..." or "it is recommended that you set..."
```suggestion
Running large numbers of rules at very short intervals can quickly clog up Task Manager throughput, leading to higher schedule drift. Use `xpack.alerting.rules.minimumScheduleInterval.value` to set a minimum schedule interval for rules. The default (and recommended) value for this configuration is `1m`. Use `xpack.alerting.rules.minimumScheduleInterval.enforce` to specify whether to strictly enforce this minimum. While the default value for this setting is `false` to maintain backwards compatibility with existing rules, set this to `true` to prevent new and updated rules from running at an interval below the minimum.
```

[float]

> ==== Rules run for a long time

Reviewer: Without "that" it seems like we're just stating in general that rules all run for a long time
```suggestion
==== Rules that run for a long time
```

> Rules that run for a long time typically do so because they are issuing resource-intensive {es} queries or performing CPU-intensive processing. This can block the event loop, making {kib} inaccessible for the duration of the rule run. By default, rule processing is cancelled after `5m` but this can be overriden using the `xpack.alerting.rules.run.timeout` configuration. This value can also be configured per rule type using `xpack.alerting.rules.run.ruleTypeOverrides`. For example, the following config sets the global timeout value to `1m` while allowing *Index Threshold* rules to run for `10m` before being cancelled.

Reviewer: Not a mandatory change, just seems like a simpler statement:
```suggestion
Rules that run for a long time typically do so because they are issuing resource-intensive {es} queries or performing CPU-intensive processing. This can block the event loop, making {kib} inaccessible while the rule runs. By default, rule processing is cancelled after `5m` but this can be overriden using the `xpack.alerting.rules.run.timeout` configuration. This value can also be configured per rule type using `xpack.alerting.rules.run.ruleTypeOverrides`. For example, the following configuration sets the global timeout value to `1m` while allowing *Index Threshold* rules to run for `10m` before being cancelled.
```

[source,yaml]
--
xpack.alerting.rules.run:
  timeout: '1m'
  ruleTypeOverrides:
    - id: '.index-threshold'
      timeout: '10m'
--

> Once a rule run is cancelled, any alerts and actions that were generated during the run are discarded. This behavior is controlled by the `xpack.alerting.cancelAlertsOnRuleTimeout` configuration, which defaults to `true`. Set this to `false` to receive alerts and actions after timeout, although be aware that these may be incomplete and possibly inaccurate.

Reviewer: IMO "after" or "when" are generally clearer than "once":
```suggestion
When a rule run is cancelled, any alerts and actions that were generated during the run are discarded. This behavior is controlled by the `xpack.alerting.cancelAlertsOnRuleTimeout` configuration, which defaults to `true`. Set this to `false` to receive alerts and actions after the timeout, although be aware that these may be incomplete and possibly inaccurate.
```

[float]

> ==== Rules spawn too many actions

Reviewer: Ditto previous section's title
```suggestion
==== Rules that spawn too many actions
```
